pageserver/src/remote_storage/storage_sync.rs

```diff
@@ -210,19 +231,134 @@ pub(super) fn spawn_storage_sync_thread<
-    let mut remote_timelines = runtime
+    let remote_timelines = runtime
         .block_on(fetch_existing_uploads(&remote_storage))
         .context("Failed to determine previously uploaded timelines")?;
 
-    let urgent_downloads = latest_timelines(&remote_timelines)
+    schedule_first_tasks(config, &remote_timelines);
+
+    let remote_timelines = Arc::new(Mutex::new(remote_timelines));
+    let remote_storage = Arc::new(remote_storage);
+
+    while !crate::tenant_mgr::shutdown_requested() {
+        let next_tasks = get_next_tasks(max_concurrent_sync);
+        let remaining_queue_length = SYNC_QUEUE.len();
+
+        if next_tasks.is_empty() {
+            trace!("No storage sync tasks found");
+            thread::sleep(Duration::from_secs(1));
+            continue;
+        } else {
+            debug!(
+                "Processing {} tasks, more tasks left to process: {}",
+                next_tasks.len(),
+                remaining_queue_length
+            );
+            REMAINING_SYNC_ITEMS.set(remaining_queue_length as i64);
+
+            let mut task_batch = next_tasks
+                .into_iter()
+                .map(|task| {
+                    tokio::spawn(process_task(
+                        config,
+                        Arc::clone(&remote_storage),
+                        task,
+                        max_sync_errors,
+                        Arc::clone(&remote_timelines),
+                    ))
+                })
+                .collect::<FuturesUnordered<_>>();
+            runtime.block_on(async move {
+                while let Some(task) = task_batch.next().await {
+                    if let Err(e) = task {
+                        error!("Failed to schedule storage sync task: {:#}", e);
+                    }
+                }
+            });
+        }
+    }
+    debug!("Queue based remote storage sync thread shut down");
+    Ok(())
+}
+
+async fn process_task<
+    P: std::fmt::Debug + Send + Sync + 'static,
+    S: RemoteStorage<StoragePath = P> + Send + Sync + 'static,
+>(
+    config: &'static PageServerConf,
+    remote_storage: Arc<S>,
+    task: SyncTask,
+    max_sync_errors: NonZeroU32,
+    remote_timelines: Arc<Mutex<HashMap<(ZTenantId, ZTimelineId), RemoteTimeline>>>,
+) {
+    if task.retries > max_sync_errors.get() {
+        error!(
+            "Evicting task {:?} that failed {} times, exceeding the error theshold",
+            task.kind, task.retries
+        );
+        let tenant_id = task.tenant_id();
+        let timeline_id = task.timeline_id();
+        if let Err(e) =
+            get_repository_for_tenant(tenant_id).and_then(|repo| repo.unload_timeline(timeline_id))
+        {
+            error!(
+                "Failed to remove repository for tenant {}, timeline {}: {:#}",
+                tenant_id, timeline_id, e
+            )
+        }
+        return;
+    }
+
+    let seconds_to_wait = f64::EPSILON.powf(task.retries.saturating_sub(1) as f64) - 1.0;
```

Comment:
Why `EPSILON`? Also I think it might be good to guard the wait behind an if for the first iteration

Reply:
Actually, wanted to use `E`, to use exponential back-off, thanks for spotting.
I did not think that if is necessary due to having `powf - 1` always >= 0, but sure, I can add one.

Reply:
I think base 2 is pretty common. You might also want to cap the maximum backoff to something reasonable (e^6 = 403 s) seems excessive. Also some recommend using a random number of milliseconds on top of this, not sure how much that matters for this use case